### PR TITLE
Baseline passage (sentence) retrieval: idf overlap #140

### LINKED ISSUE #140: Baseline passage (sentence) retrieval: idf overlap
For @rosequ 

The simplest passage (sentence) retrieval baseline is idf overlap of query terms. So I want something like:

```
sh target/appassembler/bin/RetrieveSentences -q "What is the tallest mountain in the world?"
```

The program will print out a ranked list of sentences, scored by idf overlap.

Standard pipeline architecture: it'll do doc retrieval first, fetch the raw doc text, then score each sentence.

### PR DESCRIPTION
**Pull request by rosequ**
Fixed #140. Please review.

### CODE REVIEW

File: src/main/java/io/anserini/qa/RetrieveSentences.java [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+
+    RetrieveSentences rs = new RetrieveSentences();
+    rs.retrieveDocuments(qaArgs);
+    rs.computeIDF(qaArgs);
+
+    Map<String, Double> sentences = rs.computeIDF(qaArgs);
+
+    Map.Entry<String, Double> maxEntry = null;
+    for (Map.Entry<String, Double> entry : sentences.entrySet())
+    {
```
Comment (lintool) [inline src/main/java/io/anserini/qa/RetrieveSentences.java]: braces w/ the previous line.

File: src/main/java/io/anserini/qa/RetrieveSentences.java [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+    IndexReader reader = DirectoryReader.open(FSDirectory.open(indexPath));
+
+    Directory dir;
+    dir = FSDirectory.open(Paths.get(qaArgs.index));
+
+    //using BM25 scoring model
+    Similarity similarity = null;
+    similarity = new BM25Similarity(0.9f, 0.4f);
+
+    RerankerCascade cascade = new RerankerCascade();
```
Comment (lintool) [inline src/main/java/io/anserini/qa/RetrieveSentences.java]: You're not ranking so you don't need the cascade.

Comment (rosequ) [inline src/main/java/io/anserini/qa/RetrieveSentences.java]: Oh sorry, thought I had it removed.

Comment (rosequ) [inline src/main/java/io/anserini/qa/RetrieveSentences.java]: Had a look into it: search takes cascade as an argument

**Comment by lintool** [comment]
I think we should probably build the interface for a `PassageScorer` in the `passage` package.

`PassageScorer` should have a single method - `float score(String q, String passage)` or something like that.

And Implement `TermMatchIdfPassageScorer` or something like that for your current implementation.

Later we can have `EarthMoverDistancePassageScorer`, etc.

**Comment by rosequ** [comment]
Good idea!

File: src/main/java/io/anserini/qa/RetrieveSentences.java (hunk 2 above)
Comment (lintool) [inline src/main/java/io/anserini/qa/RetrieveSentences.java]: Don't depend on `SearchWebCollection`, just copy out the lines you need. It's find to have a separate code path right now.

**Comment by lintool** [comment]
Alternatively, score might take `List<String> sentences` and keep the score internally, and then you can call `getRankedPassages` or something like that. Point is, some object needs to hold the scores in a heap...

Try working through different options and see which you like better.

**Comment by lintool** [comment]
Also, might want to normalize the idf scores by length of sentence or something like that, because otherwise you're favoring overly long sentences...

**Comment by rosequ** [comment]
Yes, I noticed that. I'll have them normalized.

**Comment by rosequ** [comment]
Please critique.

File: src/main/java/io/anserini/qa/passage/IdfPassageScorer.java [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+      for (String term: terms) {
+        try {
+          TermQuery q = (TermQuery) qp.parse(term);
+          Term t = q.getTerm();
+          idf += similarity.idf(reader.docFreq(t), reader.numDocs());
+        } catch (Exception e){
+          continue;
+        }
+      }
+      Map<String, Double> sentenceScore = new HashMap<>();
```
Comment (lintool) [inline src/main/java/io/anserini/qa/passage/IdfPassageScorer.java]: I don't like this design - see below.

File: src/main/java/io/anserini/qa/passage/PassageScorer.java [hunk 4]
```diff
@@ -0,0 +1,7 @@
+package io.anserini.qa.passage;
+
+import java.util.List;
+
+public interface PassageScorer {
+  void score(List<String> sentences, String index, String output, Context context) throws Exception;
+}
```
Comment (lintool) [inline src/main/java/io/anserini/qa/passage/PassageScorer.java]: Add method:

`List<ScoredPassage> extractTopPassages`

The internal implementation of `PassageScorer` keeps track of the top k in a heap.

Create new object: a `ScoredPassage` is just Passage (sentence String) with a score.

File: src/main/java/io/anserini/qa/RetrieveSentences.java [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+        String docid = line.trim().split(" ")[1];
+        List<Sentence> sentences = util.getSentDocument(docid);
+        for (Sentence sent : sentences) {
+          sentencesList.add(sent.text());
+        }
+      }
+    }
+    Context c = new Context();
+    scorer.score(sentencesList, qaArgs.index, qaArgs.output, c);
+    Map<String, Double> sentences = c.getState();
```
Comment (lintool) [inline src/main/java/io/anserini/qa/RetrieveSentences.java]: Get rid of this context object - the passage scorer should keep track of top k passages internally. No need for the caller to drag the scores around.

File: src/main/java/io/anserini/qa/passage/IdfPassageScorer.java [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+import org.apache.lucene.search.TermQuery;
+import org.apache.lucene.search.similarities.ClassicSimilarity;
+import org.apache.lucene.store.FSDirectory;
+
+import java.io.File;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+public class IdfPassageScorer implements PassageScorer{
```
Comment (lintool) [inline src/main/java/io/anserini/qa/passage/IdfPassageScorer.java]: space before {

File: src/main/java/io/anserini/qa/passage/IdfPassageScorer.java [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+
+  @Override
+  public void score(List<String> sentences, String index, String output, Context context) throws Exception {
+    IndexUtils util = new IndexUtils(index);
```
Comment (lintool) [inline src/main/java/io/anserini/qa/passage/IdfPassageScorer.java]: Move index initialization into a constructor for the object.

File: src/main/java/io/anserini/qa/passage/IdfPassageScorer.java [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+
+    EnglishAnalyzer ea = new EnglishAnalyzer(CharArraySet.EMPTY_SET);
+    QueryParser qp = new QueryParser(LuceneDocumentGenerator.FIELD_BODY, ea);
+
+    Map<String, Double> sentenceIDF = new HashMap();
+    ClassicSimilarity similarity = new ClassicSimilarity();
+
+    for (String sent: sentences) {
+      double idf = 0.0;
+      String[] terms = sent.split(" ");
```
Comment (lintool) [inline src/main/java/io/anserini/qa/passage/IdfPassageScorer.java]: Don't you want to tokenize the sentences also?

File: src/main/java/io/anserini/qa/passage/IdfPassageScorer.java [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+  public void score(List<String> sentences, String index, String output, Context context) throws Exception {
+    IndexUtils util = new IndexUtils(index);
+    FSDirectory directory = FSDirectory.open(new File(index).toPath());
+    DirectoryReader reader = DirectoryReader.open(directory);
+
+    EnglishAnalyzer ea = new EnglishAnalyzer(CharArraySet.EMPTY_SET);
+    QueryParser qp = new QueryParser(LuceneDocumentGenerator.FIELD_BODY, ea);
+
+    Map<String, Double> sentenceIDF = new HashMap();
+    ClassicSimilarity similarity = new ClassicSimilarity();
```
Comment (lintool) [inline src/main/java/io/anserini/qa/passage/IdfPassageScorer.java]: Why aren't we using `BM25Similarity` or `LMDirichletSimilarity`?

File: src/main/java/io/anserini/qa/RetrieveSentences.java [hunk 10]
```diff
@@ -0,0 +1,213 @@
...
+    Map.Entry<String, Double> maxEntry = null;
+    for (Map.Entry<String, Double> entry : sentences.entrySet()) {
+      if (maxEntry == null || entry.getValue().compareTo(maxEntry.getValue()) > 0) {
+        maxEntry = entry;
+      }
+    }
+    System.out.println("Answer: " + maxEntry.getKey());
+  }
+
+  public SortedMap<Integer, String> readTopicsFile(String topicsFile) throws IOException {
```
Comment (lintool) [inline src/main/java/io/anserini/qa/RetrieveSentences.java]: Refactor this out into a class, put inside `io.anserini.search.query`. Follow conventions there.

File: src/main/java/io/anserini/qa/passage/Context.java [hunk 11]
```diff
@@ -0,0 +1,19 @@
+package io.anserini.qa.passage;
+
+import java.util.Map;
+
+public class Context {
+  private Map<String, Double> sentenceScore;
+g
+  public Context(){
```
Comment (lintool) [inline src/main/java/io/anserini/qa/passage/Context.java]: typo

File: src/main/java/io/anserini/qa/passage/IdfPassageScorer.java (hunk 9 above)
Comment (rosequ) [inline src/main/java/io/anserini/qa/passage/IdfPassageScorer.java]: This is to calculate idf

File: src/main/java/io/anserini/qa/passage/IdfPassageScorer.java (hunk 8 above)
Comment (rosequ) [inline src/main/java/io/anserini/qa/passage/IdfPassageScorer.java]: The sentences will be pass through English Analyzer

File: src/main/java/io/anserini/qa/RetrieveSentences.java (hunk 5 above)
Comment (rosequ) [inline src/main/java/io/anserini/qa/RetrieveSentences.java]: I was just following State Design pattern but I'll have it changed

File: src/main/java/io/anserini/qa/passage/ScoredPassage.java [hunk 12]
```diff
@@ -0,0 +1,30 @@
...
+  double score;
+
+  public ScoredPassage(String sentence, double score) {
+    this.sentence = sentence;
+    this.score = score;
+  }
+
+  public String getSentence() {
+    return  sentence;
+  }
```
Comment (lintool) [inline src/main/java/io/anserini/qa/passage/ScoredPassage.java]: extra space

File: src/main/java/io/anserini/qa/passage/ScoredPassage.java [hunk 13]
```diff
@@ -0,0 +1,30 @@
...
+
+  public double getScore() {
+    return  score;
+  }
```
Comment (lintool) [inline src/main/java/io/anserini/qa/passage/ScoredPassage.java]: extra space

File: src/main/java/io/anserini/search/query/QaTopicReader.java [hunk 14]
```diff
@@ -0,0 +1,42 @@
...
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.List;
+import java.util.SortedMap;
+import java.util.TreeMap;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+public class QaTopicReader extends TopicReader {
```
Comment (lintool) [inline src/main/java/io/anserini/search/query/QaTopicReader.java]: This is only for the TrecQA collection, right? And it's not "topics" in the sense queries? Shouldn't it just be `TrecQAPairs` or something like that?

File: src/main/java/io/anserini/qa/passage/ScoredPassage.java (hunk 13 above)
Comment (rosequ) [inline src/main/java/io/anserini/qa/passage/ScoredPassage.java]: @lintool do you want the methods to be written single line? This convention is consistent with the one followed throughout Anserini.

Comment (lintool) [inline src/main/java/io/anserini/qa/passage/ScoredPassage.java]: nah, this is good - actually, it's IntelliJ that puts short methods on a single line...

Comment (rosequ) [inline src/main/java/io/anserini/qa/passage/ScoredPassage.java]: There is just a space separating the methods. I'll keep them?

**Pull request merged**